### ISSUE TITLE
x64-x86 Incorrect imm operand size - signed extension

### ISSUE THREAD

**Issue by UnlimitedChild** [aquynh/capstone#1092]
The size of the operand must strictly correspond to the size of the data ... 

> 		// size of this operand (in bytes).
> 		uint8_t size;

btc bx,E7
```
cstool -d x32 "66 0F BA FB E7"
0  660fbafbe7                        btc        bx, -0x19
        Prefix:0x00 0x00 0x66 0x00
        Opcode:0x0f 0xba 0x00 0x00
        rex: 0x0
        addr_size: 4
        modrm: 0xfb
        disp: 0x0
        sib: 0x0
        imm_count: 1
                imms[1]: 0xffffffffffffffe7
        op_count: 2
                operands[0].type: REG = bx
                operands[0].size: 2
                operands[1].type: IMM = 0xffffffffffffffe7
                operands[1].size: 2
```

operands[1].size: 2 - How to get the right size? Why the parser added a signed extension?

**Intel® 64 and IA-32 Architectures
Software Developer’s Manual**

Opcode | Instruction | Op/En | 64-bitMode | Compat/Leg Mode | Description
-- | -- | -- | -- | -- | --
0F BB /r | BTC r/m16, r16 | MR | Valid | Valid | Store selected bit in CF flag and complement.
0F BB /r | BTC r/m32, r32 | MR | Valid | Valid | Store selected bit in CF flag and complement.
REX.W + 0F BB /r | BTC r/m64, r64 | MR | Valid | N.E. | Store selected bit in CF flag and complement.
**0F BA /7 ib | BTC r/m16, imm8 | MI | Valid | Valid | Store selected bit in CF flag and complement.**
0F BA /7 ib | BTC r/m32, imm8 | MI | Valid | Valid | Store selected bit in CF flag and complement.
REX.W + 0F BA /7 ib | BTC r/m64, imm8 | MI | Valid | N.E. | Store selected bit in CF flag and complement

btc bx,E7
```
cstool -d x64 "66 0F BA FB E7"
0  660fbafbe7                        btc        bx, -0x19
        Prefix:0x00 0x00 0x66 0x00
        Opcode:0x0f 0xba 0x00 0x00
        rex: 0x0
        addr_size: 8
        modrm: 0xfb
        disp: 0x0
        sib: 0x0
        imm_count: 1
                imms[1]: 0xffffffffffffffe7
        op_count: 2
                operands[0].type: REG = bx
                operands[0].size: 2
                operands[1].type: IMM = 0xffffffffffffffe7
                operands[1].size: 2
```

| 6A A7                             | PUSH    FFFFFFFFFFFFFFA7                            |
| 68 A7 00 00 00               | PUSH    A7                                                     |

```
cstool -d x32 "6A A7"
0  6aa7                              push       -0x59
        Prefix:0x00 0x00 0x00 0x00
        Opcode:0x6a 0x00 0x00 0x00
        rex: 0x0
        addr_size: 4
        modrm: 0x0
        disp: 0x0
        sib: 0x0
        imm_count: 1
                imms[1]: 0xffffffffffffffa7
        op_count: 1
                operands[0].type: IMM = 0xffffffffffffffa7
                operands[0].size: 4
        Groups: not64bitmode
```

```
cstool -d x64 "6A A7"
0  6aa7                              push       0xa7
        Prefix:0x00 0x00 0x00 0x00
        Opcode:0x6a 0x00 0x00 0x00
        rex: 0x0
        addr_size: 8
        modrm: 0x0
        disp: 0x0
        sib: 0x0
        imm_count: 1
                imms[1]: 0xa7
        op_count: 1
                operands[0].type: IMM = 0xa7
                operands[0].size: 1
        Groups: mode64
```

The problem exists for a number of instructions - bt, btc, btr, bts, rcl, rcr, rol, ror, sal, sar, shl, shr, shldб shrd, probably even for other instructions with an sign extension.

 Best regards.

**Comment by condret** [comment]
https://github.com/aquynh/capstone/blob/master/arch/X86/X86Disassembler.c#L151 I think it's this function

**Comment by aquynh** [comment]
Did you try with the next branch?

**Comment by condret** [comment]
yes, and it does not work. I'm not 100% sure if I understand the documentation of these instructions correctly. I think one could just remove the sign-expansion.

**Comment by condret** [comment]
it seems like either the documentation or hardware is wrong here:
https://asciinema.org/a/6cHJj3r34xH5KtP9M1Fto9bGl

**Comment by mjhouse** [comment]
@aquynh, I think I'm having the same issue. I need the accurate operand size so that I can calculate the address of a single operand using the instruction address and the preceding operands. I'm using the next branch at commit b59bbc5b3b4999d010e6bf91499ef377d43ad0c6. As you can see in the following image, I'm iterating on operands and printing the reported size, which is always 8:

![image](https://user-images.githubusercontent.com/4141157/218842361-dcf501ea-1a9f-4a46-91a3-de27f7f41dc7.png)

This is the `objdump` output for the same assembly:

![image](https://user-images.githubusercontent.com/4141157/218842760-252f0c9b-9e4f-47c7-8e1d-7d9732b5c79c.png)
